tests/unit/test_tools.py

```diff
@@ -0,0 +1,200 @@
+"""Tests for tools.py"""
```

Comment:
Very cool testing! 
I've read through it and can follow it all, though I could not have written it myself without hundreds of hours of study. 
Looks like every aspect of deprecation is fully tested.

Reply:
The question is if it is readable enough and the only thing missing is prior knowledge. I did spend some time on it, to make it serve as an example. If not now, let me know later. Maybe we should do a small testing seminar at some point and then you can see afterwards if you can read the tests and if not I will update them. Even with test code there are questions with how far one goes to maximize code re-use at the right level, which has a tendency to make code harder to read. So that may be a problem here as well.

Reply:
I'll resolve this for now, but please re-open if you want to discuss about the readability of the test code.

Reply:
It **is** readable (in spite of also looking like magic) was my point :)